core/http/client/src/main/java/org/eclipse/rdf4j/http/client/shacl/RemoteShaclSailValidationException.java

```diff
@@ -0,0 +1,38 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+
+package org.eclipse.rdf4j.http.client.shacl;
+
+import org.eclipse.rdf4j.model.Model;
+import org.eclipse.rdf4j.model.vocabulary.SHACL;
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.eclipse.rdf4j.sail.SailException;
+import org.eclipse.rdf4j.exceptions.ValidationException;
+
+import java.io.StringReader;
+
+public class RemoteShaclSailValidationException extends SailException implements ValidationException {
+
+	private final RemoteValidation remoteValidation;
+
+	public RemoteShaclSailValidationException(StringReader stringReader, String s, RDFFormat format) {
+		remoteValidation = new RemoteValidation(stringReader, s, format);
+	}
+
+	/**
+	 * @return A Model containing the validation report as specified by the SHACL Recommendation
+	 */
+	@Override
+	public Model validationReportAsModel() {
+		Model model = remoteValidation.asModel();
+		model.setNamespace(SHACL.PREFIX, SHACL.NAMESPACE);
+		return model;
+
+	}
+
+}
```

Comment:
@jeenbroekstra 

This is the exception that gets thrown by the client when the server responds with a 409 and the validation report. The ValidationException interface is now what's common between the regular ShaclSailValidationException and this exception. Users would then need to migrate to "catch" this interface instead of the ShaclSailValidationException if they only retrieve the model and want their code to be portable between a local ShaclSail and remote one.

Things still left to consider:
 - Should this exception extend the RDF4JException (which is deprecated) or SailException (which is what the ShaclSailValidationException does? Not sure why the RDF4JException isn't an interface instead of an abstract class.
 - Should we mark this and the remote validation exception as experimental, since this isn't the final location of the interface at least?
 - How to support streaming validation reports?
 - What are we going to do with the current java based validation report? Maybe deprecate it so we can move it and generalise it?

Reply:
My take on things: 
- RDF4JException isn't deprecated. It inherits from the old OpenRDFException (which _is_ deprecated, and is a class because we made a design decision that all RDF4J exceptions are unchecked (so it needs to extend RuntimeException). Actually SailException itself extends RDF4JException as well (_all_ RDF4J exceptions do).  I still find the naming not quite right. The word "SAIL" shouldn't appear on the client side at all. Sails are an internal server-side matter. So the server-side version of the ValidationException should be called ShaclValidationException or something like that, and probably extend SailException. But the client-side version should be called something like RemoteShaclValidationException or something like that, but _not_ have Sail in its name, and not extend SailException
- marking things experimental is always good for a new feature - it gives us leeway :)
- YAGNI (or at the very least, not in scope for this iteration, though I'm fairly confident we can do something along these lines when it becomes necessary)
- is this the report functionality as previously released you mean? If we can deprecate it, keep it functional but then copy and migrate, that'd be ideal.